src/environment.rs

```diff
@@ -0,0 +1,124 @@
+    /// initialized with abstract_values::Top
+    pub fn with_parameters(def_id: DefId, tcx: TyCtxt) -> Environment {
+        let value_map = HashTrieMap::default();
+        let func_ty: Ty = tcx.type_of(def_id);
+        match func_ty.sty {
+            ty::FnDef(..) | ty::FnPtr(_) => {
+                let sig = func_ty.fn_sig(tcx);
+                let len = sig.inputs().skip_binder().len();
+                for i in 0..len {
+                    let par_i = Path::Parameter { index: i };
+                    // todo: figure out how to get a source span for each parameter definition
+                    value_map.insert(par_i, abstract_value::TOP);
+                }
+            }
+            _ => {}
+        };
+        Environment { value_map }
+    }
+}
+
+/// Methods
+impl Environment {
+    /// Returns a reference to the value associated with the given path.
+    /// If no such value exists, &abstract_value::Bottom is returned.
+    pub fn value_at(&self, path: &Path) -> &AbstractValue {
+        self.value_map.get(path).unwrap_or(&abstract_value::BOTTOM)
+    }
+
+    /// Returns an environment with a path for every entry in self and other and an associated
+    /// value that is the join of self.value_at(path) and other.value_at(path)
+    pub fn join(&self, other: &Environment, join_condition: &AbstractValue) -> Environment {
+        self.join_or_widen(other, join_condition, |x, y, c| x.join(y, c))
+    }
+
+    /// Returns an environment with a path for every entry in self and other and an associated
+    /// value that is the widen of self.value_at(path) and other.value_at(path)
+    pub fn widen(&self, other: &Environment, join_condition: &AbstractValue) -> Environment {
+        self.join_or_widen(other, join_condition, |x, y, c| x.widen(y, c))
+    }
+
+    /// Returns an environment with a path for every entry in self and other and an associated
+    /// value that is the join or widen of self.value_at(path) and other.value_at(path).
+    fn join_or_widen(
+        &self,
+        other: &Environment,
+        join_condition: &AbstractValue,
+        join_or_widen: fn(&AbstractValue, &AbstractValue, &AbstractValue) -> AbstractValue,
+    ) -> Environment {
+        let value_map1 = &self.value_map;
+        let value_map2 = &other.value_map;
+        let mut value_map: HashTrieMap<Path, AbstractValue> = HashTrieMap::default();
+        for (path, val1) in value_map1.iter() {
+            let p = path.clone();
+            match value_map2.get(path) {
+                Some(val2) => {
+                    value_map = value_map.insert(p, join_or_widen(&val1, &val2, &join_condition));
+                }
+                None => {
+                    debug_assert!(!val1.is_bottom());
+                    value_map = value_map.insert(
+                        p,
+                        join_or_widen(&val1, &abstract_value::BOTTOM, &join_condition),
```

Comment:
Would using `val1` yield a different result? It seems like it would be equivalent except in the case of a super-aggressive widening that goes directly to Top...

Reply:
This is a bit confusing. If we did not keep track of the join condition, val1 would do just nicely. It is not totally clear to me right now that it keeping the join condition will matter, but in general I prefer to keep as much information around as possible, for use by the expression domain.